[Dynamo] Support methods of NamedTuple

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103217

This PR adds the support of calling NamedTuple methods. https://github.com/pytorch/pytorch/issues/91662

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

fegin committed:
[Dynamo] Support methods of NamedTuple

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/lists.py:
```diff
@@ -547,9 +549,27 @@ def reconstruct(self, codegen):
+                    method.__func__, variables.UserDefinedClassVariable(self.tuple_cls)
+                )
+            elif inspect.isfunction(method):
+                if "self" in inspect.signature(method).parameters:
```
Is there not a better way to check if `method` is a static method or normal method other than checking for the name of the argument? There is no guarantee that `self` is the name of the first argument for normal methods.

Can we use `method = inspect.getattr_static(self.tuple_cls, name, None)` and `if isinstance(method, [staticmethod/classmethod])`?

wconstab replied:
in addition to the above, I am wondering if we can call a helper like VariableBuilder()(method) to produce the correct wrapping.  That way the logic for checking types/args is in one place.

That being said, I have written/seen code like this in a lot of other VariableTrackers, and there might be a reason doing it in VariableBuilder isn't workable.

fegin replied:
@williamwen42, sounds good. The original rationale is to get `FunctionType` with `gettattr` as `UserFunctionVariable` does not accept `staticmethod`. But I realize I can use `__func__`.  Will change to use `inspect.getattr_static`.

jansel commented on torch/_dynamo/variables/lists.py:
```diff
@@ -547,9 +549,27 @@ def reconstruct(self, codegen):
 
     def var_getattr(self, tx, name):
+        def check_and_create_method():
+            method = getattr(self.tuple_cls, name, None)
```
use inspect.getattr_static to not run arbitrary code

fegin committed:
Update on "[Dynamo] Support methods of NamedTuple"


This PR adds the support of calling NamedTuple methods. https://github.com/pytorch/pytorch/issues/91662

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

fegin committed:
Update on "[Dynamo] Support methods of NamedTuple"


This PR adds the support of calling NamedTuple methods. https://github.com/pytorch/pytorch/issues/91662

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge
